Initial version of pyston_lite

kmod opened the pull request:
This is only a few percent faster, but I wanted to put up an initial PR and will add some optimizations in later PRs.

I originally did the development by copying all the jit files into a separate directory, but this made for a tough merge. I think it makes sense to try to keep just a single version of our jit with appropriate ifdefs for pyston_lite, so that's how this PR does it.

This PR is just the basics to get pyston_lite working, which is mainly removing the features that we can't use, and finding replacements for the ones that we still need. From a perf standpoint the biggest hit is the removal of the split-dict caches, but I will add my new version of the cache that helps with that.

In a future PR I will also make this a bit more safe in case you run for a while before switching to pyston_lite (in particular, dealing with any opcaches that were already initialized).

I used a bare-bones Makefile for now; you can test it by going into `pyston/pyston_lite` and doing `make bench` or `make bench_baseline`. I'll migrate this to a setup.py in a later PR

kmod committed:
pyston-lite initial commit: symlink in some files and add a Makefile

Not functional yet

kmod committed:
update include paths

kmod committed:
Enable opcache stats from aot_ceval_jit_helper.c

they were previously always-off

kmod committed:
All the changes to get pyston_lite working

- Remove caches that use the split-dict keys version that we added
- Disable AOT entirely
- Have to do some trickery to get access to static functions inside cpython
- Have to use PyCode_GetExtra instead of adding a field to PyCodeObject
- Add our own custom opcache
- Make it into a module that can be imported and enabled
- I added a number of functions to lib.c that we put into other places
- I renamed some of the publicly-available function names so that they don't clash with pyston-full, in case you try to load pyston-lite into pyston full (though I haven't tested that in a while)

kmod committed:
Disable definedness optimizations

I changed the implementation so that it completely compiles out support for it, just to be extra sure that we're not mistakenly using uninitialized data.

kmod committed:
.pth file that automatically loads pyston-lite

I'm imagining that we will offer a "pyston_lite" package which you
can enable manually, and a "pyston_lite_autoload" package
which will load pyston_lite automatically.

Currently it loads pyston_lite if you set the PYSTON env var

kmod committed:
pyston_lite currently crashes when loaded into pyston

not sure why, but in the past it was due to name conflicts between the two
and ending up using some functions from pyston_lite and some from pyston_full

kmod commented:
I just noticed a crash when running the cpython testsuite, I'll take a look into it

kmod committed:
makefile target to run the testsuite'

kmod committed:
fix: use the cpython version of this file

We want to use the Pyston version of pycore_code.h, since we get
to control the opcache and use our version of that file.

But for dict-common.h, we want to use the cpython version because
we have to use their object layout.

kmod committed:
Run the non-buggy parts of test_code

kmod commented:
ok I fixed the crash and this is ready for review

undingen commented on pyston/pyston_lite/aot_ceval_jit_helper.h:
```diff
@@ -0,0 +1 @@
+../../Python/aot_ceval_jit_helper.h
```
This file does not exist

undingen commented on Python/aot_ceval.c:
```diff
@@ -1838,30 +1943,52 @@ _PyEval_EvalFrame_AOT_Interpreter(PyFrameObject *f, int throwflag, PyThreadState
+    if (can_use_jit && co->co_opcache_flag >= jit_min_runs /* jit after that many calls or gen yields */) {
+#ifdef PYSTON_LITE
+        void* code = JIT_FUNC_FAILED;
+        _PyCode_GetExtra((PyObject*)co, code_jitfunc_index, &code);
```
I find all this ifdefs and code duplication just to set and retrieve the the pointer to the machine code a bit ugly. What do you think about adding a helper function/macro which handles it?!?

undingen commented on Python/aot_ceval_jit.c:
```diff
@@ -847,7 +900,11 @@ static void switch_section(Jit* Dst, Section new_section) {
 |.macro type_version_check, r_type_idx, type_ver, false_branch
-|| emit_cmp64_mem_imm(Dst, r_type_idx, offsetof(PyTypeObject, tp_version_tag), (uint64_t)type_ver);
+#ifdef PYSTON_LITE
+|| emit_cmp32_mem_imm(Dst, r_type_idx, offsetof(PyTypeObject, tp_version_tag), (unsigned int)type_ver);
```
I had totally forgot about the difference in size of `tp_version_tag`, good that you noticed!
But the #ifdef/else/endif needs "||" infront. Right now dynasm generates both paths because it just seems to ignore them:
```C
    } else {
        // PyTypeObject *arg2 = Py_TYPE(obj)
        emit_load64_mem(Dst, arg2_idx, arg1_idx,  offsetof(PyObject, ob_type));
        //| type_version_check, arg2_idx, la->type_ver, >1
        emit_cmp32_mem_imm(Dst, arg2_idx, offsetof(PyTypeObject, tp_version_tag), (unsigned int)la->type_ver);
        emit_cmp64_mem_imm(Dst, arg2_idx, offsetof(PyTypeObject, tp_version_tag), (unsigned int)la->type_ver);
        dasm_put(Dst, 323);
```

undingen commented on pyston/pyston_lite/pyston_lite_autoload.pth:
```diff
@@ -0,0 +1 @@
+import os, sys; exec("""sys.path.append("/home/kmod/pyston/pyston/pyston_lite"); ("PYSTON" in os.environ) and __import__("pyston_lite").enable()""")
```
hard coded user `kmod` in path. Would be nice if this file would get generated or you find another way to get the path.

undingen commented on pyston/pyston_lite/Makefile:
```diff
@@ -51,6 +51,7 @@ BENCH:=../macrobenchmarks/benchmarks/bm_flaskblogging/run_benchmark.py --legacy
 # note: the testsuite doesn't pass for my Ubuntu system cpython, but it does pass
 # if I reconfigure this makefile to use a manually-built cpython
 testsuite: env
+	PYSTON=1 ./env/bin/python -c 'import test.support; test.support.check_impl_detail = lambda **kw: False; import test.test_code; test.test_code.test_main()'
```
It can't find test_code for me not sure how I have to generate it.

kmod replied:
I had to do `sudo apt-get install libpython3.8-testsuite` since I think Ubuntu strips out the tests from their package. If you test this against a manually-built cpython I think it just works.

undingen replied:
a thanks that fixed it

undingen reviewed (commented):
Cool new era starting I'm excited about the possibilities :sparkles:. I think it would also be easy to get `pyston_lite` working on macos (as long as we stick to gcc) later on.

LGTM
I agree with sharing as much code as possible and adding ifdefs looks like the best way to proceed.
Added a few smaller comments

undingen commented:
I just noticed some problems: I think we have to check `type->tp_flags &Py_TPFLAGS_VALID_VERSION_TAG` in every IC because stock
`PyType_Modified` does not set `tp_version_tag` to 0 :(.
Also there is a problem with it overflowing: 3.8 will just clear the internal `method_cache` and reassign the numbers which I think we can't really fix just have to hope 32bit is enough :(.

kmod committed:
Reduce duplication by making some helper functions

kmod committed:
whoops linked this to the wrong place

it's not even necessary in this commit because it adds
enough include paths that it can find the original, but
in the later setup.py change this becomes an issue.

kmod committed:
review comment: dynasm apparently just ignores these

kmod committed:
whoops, left this testing code in

kmod committed:
use our objdump script to get jit disassemblies

kmod committed:
add some #line directives to the dynasm preprocessor

I'm seeing gdb display the wrong line numbers for things, and I think
what's going on is that #line directives inside a non-taken #if block are
ignored.

So when dynasm inserts some lines inside an #if block as well as a #line
directive, the #line directive is ignored but the extra lines are still counted,
resulting in too-large of a line number.

So this change adds extra #line directives after preprocessor directives.

My understanding is this issue was present pre-pyston-lite, we just didn't
see it until pyston-lite added a number of #if blocks.

It also adds line numbers after any dynasm directives so that all #line directives
point at the original file -- for consistency it's nice if they point to the same one,
and I think it's nicer if they point to the original so that you can do things like
"break aot_ceval_jit.c:2408" instead of "break aot_ceval_jit.prep.c:2408"

kmod committed:
pyston-lite: check for VALID_VERSION_TAG

cpython pre-3.10 doesn't clear tp_version_tag when the type
is modified, so we need to explicitly check for this other bit.

kmod commented:
Ok I pushed a bunch of new commits, would you mind taking a look?

undingen commented:
LGTM.

undingen commented on Python/aot_ceval_jit.c:
```diff
@@ -915,6 +915,11 @@ static void switch_section(Jit* Dst, Section new_section) {
 ||#ifdef PYSTON_LITE
+|| JIT_ASSERT(Py_TPFLAGS_VALID_VERSION_TAG == (1UL << 19), "need to update these offsets");
+@ARM|| emit_load32_mem(Dst, get_tmp_reg(r_type_idx), r_type_idx, offsetof(PyTypeObject, tp_flags))
+@ARM|  tbz get_tmp_reg(r_type_idx), 19, false_branch // Test Bit Zero: jump to false_branch if bit 19 is zero
```
`@ARM|  tbz Rw(get_tmp_reg(r_type_idx)), #19, false_branch`

undingen replied:
There are two places which reference `PyObject_IsTrueProfile` on arm64 I think it's safe to change them to `PyObject_IsTrue`

undingen commented on Python/aot_ceval_jit.c:
```diff
@@ -915,6 +915,11 @@ static void switch_section(Jit* Dst, Section new_section) {
 |.macro type_version_check, r_type_idx, type_ver, false_branch
 ||#ifdef PYSTON_LITE
+|| JIT_ASSERT(Py_TPFLAGS_VALID_VERSION_TAG == (1UL << 19), "need to update these offsets");
+@ARM|| emit_load32_mem(Dst, get_tmp_reg(r_type_idx), r_type_idx, offsetof(PyTypeObject, tp_flags))
```
missing ;

undingen reviewed (commented):
found a few small problems on arm

kmod committed:
arm fixes
